File: python/leet_code/85.py
import json
import sys
from typing import List

def find_maximal_rectangle(matrix: List[List[str]]) -> int:
    """
    This function finds the maximal rectangle in a binary matrix.

    Args:
        matrix (List[List[str]]): A binary matrix represented as a list of lists.

    Returns:
        int: The area of the maximal rectangle.

    Examples:
        >>> find_maximal_rectangle([['1', '0', '1', '0', '0'], ['1', '0', '1', '1', '1'], ['1', '1', '1', '1', '1'], ['1', '0', '0', '1', '0']])
        6
        >>> find_maximal_rectangle([['0']])
        0
        >>> find_maximal_rectangle([['1']])
        1
        >>> find_maximal_rectangle([['1', '1', '1', '1'], ['1', '1', '1', '1'], ['1', '1', '1', '1']])
        12
        >>> find_maximal_rectangle([['1', '0', '1', '0'], ['1', '0', '1', '1'], ['1', '1', '1', '1'], ['1', '0', '1', '0']])
        4
    """
    m = len(matrix)
    n = len(matrix[0])
    heights = [0] * (n + 1)
    best = 0
    for row in matrix:
        for col in range(n):
            # Update the height of the current column
            heights[col] = heights[col] + 1 if row[col] == '1' else 0
        stack = [-1]
        for col in range(n + 1):
            # Pop the stack while the current height is less than the last one in the stack
            while heights[col] < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = col - stack[-1] - 1
                # Update the best area
                best = max(best, h * w)
            stack.append(col)
    return best


if __name__ == '__main__':
    with open('user.out', 'w') as f:
        for matrix in map(json.loads, sys.stdin):
            result = find_maximal_rectangle(matrix)
            print(result, file=f)
    sys.exit()

File: rust/leet_code/src/85.rs
use std::cmp::max;

/// This module provides a solution for the problem of finding the maximal rectangle in a matrix.
/// The matrix is represented as a vector of vectors of characters, where '1' represents a filled cell and '0' represents an empty cell.
/// The function `maximal_rectangle` finds the largest rectangle consisting of '1's in the given matrix and returns its area.

impl Solution {
    /// Finds the maximal rectangle in a matrix.
    ///
    /// # Arguments
    ///
    /// * `matrix` - A vector of vectors of characters representing the matrix.
    ///
    /// # Returns
    ///
    /// * An i32 representing the area of the maximal rectangle.
    pub fn maximal_rectangle(matrix: Vec<Vec<char>>) -> i32 {
        // Initialize the height vector with zeros
        let mut h: Vec<i32> = vec![0; matrix[0].len()];
        // Initialize the result
        let mut res = 0;
        // Iterate over the rows of the matrix
        for i in 0..matrix.len() {
            // Initialize the stack
            let mut v: Vec<usize> = vec![];
            // Iterate over the columns of the matrix
            for j in 0..matrix[i].len() {
                // Update the height vector
                h[j] = if matrix[i][j] == '0' { 0 } else { h[j] + 1 };
                // Process the stack
                loop {
                    match v.last_mut() {
                        // If the current height is less than the height at the top of the stack
                        Some(&mut k) if h[k] > h[j] => {
                            v.pop();
                            res = max(
                                res,
                                h[k] * (j as i32 - if v.is_empty() { -1 } else { *v.last().unwrap() as i32 } - 1),
                            );
                        }
                        // If the current height is equal to the height at the top of the stack
                        Some(p) if h[*p] == h[j] => {
                            *p = j;
                            break;
                        }
                        // If the stack is empty or the current height is greater than the height at the top of the stack
                        _ => {
                            v.push(j);
                            break;
                        }
                    }
                }
            }
            // Process the remaining elements in the stack
            while let Some(k) = v.pop() {
                res = max(
                    res,
                    h[k] * (matrix[i].len() as i32 - if v.is_empty() { -1 } else { *v.last().unwrap() as i32 } - 1),
                );
            }
        }
        // Return the result
        res
    }
}
